File: lib/analyzer/applications/python/pip.ts
import { DepGraph, DepGraphBuilder } from "@snyk/dep-graph";
import * as Debug from "debug";
import { eventLoopSpinner } from "event-loop-spinner";
import * as path from "path";
import * as semver from "semver";
import { DepGraphFact } from "../../../facts";
import { compareVersions } from "../../../python-parser/common";
import { getPackageInfo } from "../../../python-parser/metadata-parser";
import { getRequirements } from "../../../python-parser/requirements-parser";
import {
  PythonMetadataFiles,
  PythonPackage,
  PythonRequirement,
} from "../../../python-parser/types";
import { AppDepsScanResultWithoutTarget, FilePathToContent } from "../types";

const debug = Debug("snyk");
class PythonDepGraphBuilder {
  private requirements: PythonRequirement[];
  private metadata: PythonMetadataFiles;
  private builder: DepGraphBuilder;
  private visitedMap: Set<string> = new Set();

  constructor(
    name: string,
    requirements: PythonRequirement[],
    metadata: PythonMetadataFiles,
  ) {
    this.requirements = requirements;
    this.metadata = metadata;
    this.builder = new DepGraphBuilder({ name: "pip" }, { name });
  }

  public async build(): Promise<DepGraph> {
    for (const dep of this.requirements) {
      await this.addDependenciesToDepGraph(this.builder.rootNodeId, dep);
    }
    return this.builder.build();
  }

  // depth-first search for dependencies and assigning them to the dep graph builder
  private async addDependenciesToDepGraph(
    root: string,
    req: PythonRequirement,
  ): Promise<void> {
    if (eventLoopSpinner.isStarving()) {
      await eventLoopSpinner.spin();
    }
    const metadata = this.findMetadata(req);
    if (!metadata) {
      return;
    }
    const extrasId = req.extras?.length ? `:${req.extras}` : "";
    const nodeId = `${metadata.name}@${metadata.version}${extrasId}`;
    if (!this.visitedMap.has(nodeId)) {
      this.visitedMap.add(nodeId);
      this.builder.addPkgNode(
        { name: metadata.name, version: metadata.version },
        nodeId,
      );
      for (const dep of metadata.dependencies) {
        if (this.shouldTraverse(req, dep)) {
          await this.addDependenciesToDepGraph(nodeId, dep);
        }
      }
    }
    this.builder.connectDep(root, nodeId);
  }

  // test extras and environment markers to determine whether a dependency is optional
  // if it is optional only traverse if the requirement asked for those optionals
  private shouldTraverse(
    req: PythonRequirement,
    dep: PythonRequirement,
  ): boolean {
    // always traverse deps with no extra environment markers (they're non-optional)
    if (!dep.extraEnvMarkers || dep.extraEnvMarkers.length === 0) {
      return true;
    }

    // determine if dep was required with extras, and those extras match the deps env markers
    const intersection = req.extras?.filter((i) =>
      dep.extraEnvMarkers?.includes(i),
    );

    // yes! this is an optional dependency that was asked for
    if (intersection && intersection.length > 0) {
      return true;
    }

    return false; // no! stop here we don't want to traverse optional dependencies
  }

  // find the best match for a dependency in found metadata files
  private findMetadata(dep: PythonRequirement): PythonPackage | null {
    const nameMatches = this.metadata[dep.name.toLowerCase()];
    if (!nameMatches || nameMatches.length === 0) {
      return null;
    }
    if (nameMatches.length === 1 || !dep.version) {
      return nameMatches[0];
    }
    for (const meta of nameMatches) {
      if (
        semver.satisfies(meta.version, `${dep.specifier}${dep.version}`, true)
      ) {
        return meta;
      }
    }
    // fallback to the first metadata file if no match is found
    return nameMatches[0];
  }
}

/**
 * Creates a dep graph for every requirements.txt file that was found
 */
export async function pipFilesToScannedProjects(
  filePathToContent: FilePathToContent,
  collectApplicationFiles: boolean,
): Promise<AppDepsScanResultWithoutTarget[]> {
  const scanResults: AppDepsScanResultWithoutTarget[] = [];
  const requirements = {};
  const metadataItems: PythonMetadataFiles = {};

  const filePaths = Object.keys(filePathToContent);
  for (const filepath of filePaths) {
    const fileBaseName = path.basename(filepath);
    if (fileBaseName === "requirements.txt") {
      requirements[filepath] = getRequirements(filePathToContent[filepath]);
    } else if (fileBaseName === "METADATA") {
      try {
        const packageInfo = getPackageInfo(filePathToContent[filepath]);
        if (!metadataItems[packageInfo.name.toLowerCase()]) {
          metadataItems[packageInfo.name.toLowerCase()] = [];
        }
        metadataItems[packageInfo.name.toLowerCase()].push(packageInfo);
      } catch (err) {
        debug(err.message);
      }
    }
  }
  if (Object.keys(metadataItems).length === 0) {
    return scanResults;
  }
  // pre-sort each package name by version, descending
  for (const name of Object.keys(metadataItems)) {
    metadataItems[name].sort((v1, v2) => {
      return compareVersions(v1.version, v2.version);
    });
  }

  const requirementsFiles = Object.keys(requirements);
  for (const requirementsFile of requirementsFiles) {
    if (requirements[requirementsFile].length === 0) {
      continue;
    }
    const builder = new PythonDepGraphBuilder(
      requirementsFile,
      requirements[requirementsFile],
      metadataItems,
    );
    const depGraph = await builder.build();
    if (!depGraph) {
      continue;
    }

    const depGraphFact: DepGraphFact = {
      type: "depGraph",
      data: depGraph,
    };
    scanResults.push({
      facts: [depGraphFact],
      identity: {
        type: "pip",
        targetFile: requirementsFile,
      },
    });
  }

  return scanResults;
}

File: go/pkg/analyzer/applications/python/pip.go
// Package python provides Python application dependency scanning.
// It mirrors lib/analyzer/applications/python/pip.ts.
package python

import (
	"path/filepath"
	"strings"

	"github.com/snyk/snyk-docker-plugin/pkg/depgraph"
	"github.com/snyk/snyk-docker-plugin/pkg/pythonparser"
	"github.com/snyk/snyk-docker-plugin/pkg/types"
)

// AppScanResult is a single per-file application scan result.
type AppScanResult struct {
	Identity types.Identity
	Facts    []types.Fact
}

// ScanPip builds dep-graphs for all requirements.txt + site-packages combos.
// pathToContent is a map of absolute paths to raw file bytes.
// It returns one AppScanResult per requirements.txt that has at least one dep.
func ScanPip(pathToContent map[string][]byte) []AppScanResult {
	// Separate requirements files from dist-info METADATA files.
	requirementsFiles := map[string][]byte{}
	for path, content := range pathToContent {
		if filepath.Base(path) == "requirements.txt" {
			requirementsFiles[path] = content
		}
	}

	// Parse all METADATA files (from all site-packages directories in layers).
	metadata := pythonparser.ParseSitePackagesMetadata(pathToContent)

	// Without metadata we cannot build dep-graphs (mirrors TS behaviour).
	if len(metadata) == 0 {
		return nil
	}

	var results []AppScanResult
	for _, entry := range requireirementsFilesSorted(requirementsFiles) {
		reqPath, reqContent := entry.path, entry.content
		reqs, err := pythonparser.ParseRequirementsTxt(string(reqContent))
		if err != nil || len(reqs) == 0 {
			continue
		}
		dg := buildDepGraph(reqPath, reqs, metadata)
		if dg == nil {
			continue
		}
		results = append(results, AppScanResult{
			Identity: types.Identity{
				Type:       "pip",
				TargetFile: reqPath,
			},
			Facts: []types.Fact{
				{Type: types.FactDepGraph, Data: *dg},
			},
		})
	}
	return results
}

// requirementsFilesSorted returns requirements files in deterministic order.
func requireirementsFilesSorted(m map[string][]byte) []requirementsEntry {
	out := make([]requirementsEntry, 0, len(m))
	for k, v := range m {
		out = append(out, requirementsEntry{k, v})
	}
	// Sort by path for deterministic output.
	for i := 1; i < len(out); i++ {
		for j := i; j > 0 && out[j].path < out[j-1].path; j-- {
			out[j], out[j-1] = out[j-1], out[j]
		}
	}
	return out
}

type requirementsEntry struct {
	path    string
	content []byte
}

// buildDepGraph builds a DepGraphData for a single requirements file.
func buildDepGraph(
	reqPath string,
	reqs []pythonparser.Requirement,
	metadata pythonparser.PythonMetadataFiles,
) *types.DepGraphData {
	visited := map[string]bool{}
	deps := resolveRequirements(reqs, metadata, visited, 0)
	if len(deps) == 0 {
		return nil
	}
	dg := depgraph.FromDepTree("pip", reqPath, "", deps)
	return &dg
}

// resolveRequirements resolves a list of requirements to DepInfo entries,
// following transitive dependencies. maxDepth prevents infinite recursion.
const maxPipDepth = 10

func resolveRequirements(
	reqs []pythonparser.Requirement,
	metadata pythonparser.PythonMetadataFiles,
	visited map[string]bool,
	depth int,
) []depgraph.DepInfo {
	if depth > maxPipDepth {
		return nil
	}
	var result []depgraph.DepInfo
	for _, req := range reqs {
		pkg := findPackage(req, metadata)
		if pkg == nil {
			continue
		}
		nodeKey := pkg.Name + "@" + pkg.Version
		if visited[nodeKey] {
			continue
		}
		visited[nodeKey] = true
		// Only traverse non-extra-gated deps.
		var transitiveDeps []pythonparser.Requirement
		for _, d := range pkg.Dependencies {
			if !hasExtraMarker(d, req.Extras) {
				transitiveDeps = append(transitiveDeps, d)
			}
		}
		children := resolveRequirements(transitiveDeps, metadata, visited, depth+1)
		result = append(result, depgraph.DepInfo{
			Name:    pkg.Name,
			Version: pkg.Version,
			Deps:    children,
		})
	}
	return result
}

// hasExtraMarker returns true if the requirement's extras intersect the
// requestedExtras — i.e. we should NOT traverse this dep (it's optional).
// Wait — inverted: if the dep has no extra env markers it IS traversed;
// if it has extras but none match, skip it.
// Actually we want to skip deps that require extras we didn't ask for.
// This mirrors TS shouldTraverse().
func hasExtraMarker(dep pythonparser.Requirement, requestedExtras []string) bool {
	// If the dep has no extras it is always included — return false (don't skip).
	if len(dep.Extras) == 0 {
		return false
	}
	// If the parent req requested those extras, include them.
	for _, de := range dep.Extras {
		for _, re := range requestedExtras {
			if strings.EqualFold(de, re) {
				return false // include
			}
		}
	}
	return true // skip
}

// findPackage finds the best-matching package for a requirement in metadata.
func findPackage(req pythonparser.Requirement, metadata pythonparser.PythonMetadataFiles) *pythonparser.PythonPackage {
	matches := metadata[req.Name]
	if len(matches) == 0 {
		return nil
	}
	if len(matches) == 1 || req.Version == "" {
		return &matches[0]
	}
	// Try to find the version that satisfies the specifier.
	for i := range matches {
		if versionSatisfies(matches[i].Version, req.Specifier, req.Version) {
			return &matches[i]
		}
	}
	// Fallback to first.
	return &matches[0]
}

// versionSatisfies is a simple version comparator sufficient for pip specifiers.
// Full PEP 440 semantics are not required — just string equality for ==,
// and lexicographic comparisons for >=, <=, >, <.
func versionSatisfies(have string, spec pythonparser.Specifier, want string) bool {
	switch spec {
	case pythonparser.SpecEq, pythonparser.SpecAEq:
		return have == want
	case pythonparser.SpecNe:
		return have != want
	case pythonparser.SpecGte:
		return have >= want
	case pythonparser.SpecLte:
		return have <= want
	case pythonparser.SpecGt:
		return have > want
	case pythonparser.SpecLt:
		return have < want
	case pythonparser.SpecCom:
		// ~= X.Y means >= X.Y, == X.* — approximate, use >= for MVP.
		return have >= want
	}
	return true
}
